File: leetcode_go/146.LRU_Cache.go
// https://leetcode.com/problems/lru-cache/

import "container/list"

type LRUCache struct {
	key2node map[int]*list.Element
	key2val  map[int]int
	nodes    *list.List
	capacity int
}

func Constructor(capacity int) LRUCache {
	return LRUCache{
		capacity: capacity,
		nodes:    list.New(),
		key2node: make(map[int]*list.Element, capacity),
		key2val:  make(map[int]int),
	}
}

func (this *LRUCache) Get(key int) int {
	if node, ok := this.key2node[key]; !ok {
		return -1
	} else {
		val := this.key2val[key]
		this.nodes.MoveToBack(node)
		return val
	}
}

func (this *LRUCache) Put(key int, value int) {
	if this.Get(key) != -1 {
		this.key2val[key] = value
		return
	} else {
		if len(this.key2node) == this.capacity {
			toDelete := this.nodes.Front()
			this.nodes.Remove(toDelete)
			delete(this.key2node, toDelete.Value.(int)) //Type assertions
			delete(this.key2val, toDelete.Value.(int))  //Type assertions
		}

		this.nodes.PushBack(key)
		this.key2node[key] = this.nodes.Back()
		this.key2val[key] = value
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: leetcode_cpp/146.LRU_Cache.cpp
// https://leetcode.com/problems/lru-cache/

class LRUCache {
 private:
  unordered_map<int, int> key2val;
  list<int> nodes;
  int cap;
  unordered_map<int, list<int>::iterator> key2node;

 public:
  LRUCache(int capacity) { cap = capacity; }

  int get(int key) {
    if (key2node.find(key) == key2node.end()) {
      return -1;
    }
    auto node = key2node[key];
    nodes.erase(node);
    nodes.push_back(key);
    key2node[key] = --nodes.end();
    return key2val[key];
  }

  void put(int key, int value) {
    if (get(key) != -1) {
      key2val[key] = value;
      return;
    }

    if (key2node.size() == cap) {
      auto to_delete = nodes.begin();
      key2val.erase(*to_delete);
      key2node.erase(*to_delete);
      nodes.erase(to_delete);
    }

    nodes.push_back(key);
    key2val[key] = value;
    key2node[key] = --nodes.end();
  }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */